Enh/quantum experiment plotting

nathlacroix opened the pull request:
Allows to plot all sequences through a convenience function of the quantum experiment.
Inviting @chellings to review.

nathlacroix committed:
Add saving kwargs to segment.plot

nathlacroix committed:
Add wrapper plotting of sequence to QExp

nathlacroix committed:
Add docstring for qexp.plot

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        sequences), an integer (index of sequence to plot),  or a list of
+        integers/str. If strings are in the list, then plots only sequences
+        with the corresponding name.
+        :param segments (int, list): index of segments to plot. Plots all
```
"index or list of indices"? or does it need to be a list of lists (segs for each seq)?

nathlacroix replied:
clarified

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+                        #  time we reload qexp, we would have to reset the
+                        #  datafolder... For now, current implementation below
+                        #  will work if a_tools is already loaded.
+                        folder = a_tools.get_folder(timestamp=self.timestamp)
```
- You should not access a global with a hard coded-name here. Even it a_tools is loaded in the init script, the users should be allowed to use a different name for it.
- I think you can (and should) import a_tools here (either locally or in the header of the module, but I would do it locally if you only need it here). This will not trigger a reloading of a_tools if a_tools has already been loaded before (in the init or any other module, and regardless of the alias name used previously). I do not think that you need to reset the a_tools.data_dir after reloading the qe module since a_tools will not be reloaded in that case. Please tell me if I am wrong.
- You should use the data folder of MC and not of a_tools, i.e., call a_tools.data_from_time(timestamp, folder=self.MC.datadir(), auto_fetch=False). We might consider moving (part of) the functionality of data_from_time to utilities.general since it is not specific to a_tools provided that we make the folder mandatory. a_tools can then still provide a wrapper that makes the folder optional when called via a_tools.
- Let's discuss orally if my comments are unclear.

nathlacroix replied:
done

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+                                    'experiment. Sequence plot will be saved in '
+                                    'current directory.')
+                        folder = ""
+                    save_kwargs = dict(fname= folder + "/" +
```
- you should use os.path.join
- in case of folder == "", this will lead to a leading slash, i.e., the figure will not be save in the current directory, but in the root folder

nathlacroix replied:
done

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        :param segments (int, list): index of segments to plot. Plots all
+        segments by default.
+        :param qubits (list): list of qubits to plot
+        :param save (bool): save the figures.
```
please add info about where it will be saved

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        with the corresponding name.
+        :param segments (int, list): index of segments to plot. Plots all
+        segments by default.
+        :param qubits (list): list of qubits to plot
```
- please add default behavior
- please clarify whether list of objects or list of strings (names). Consider passing it to self.get_qubits to allow the user passing qb names.

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        if qubits is None:
+            qubits = self.meas_objs
+        plot_kwargs.update(dict(channel_map=plot_kwargs.pop('channel_map',
+                           general.get_channel_map(qubits))))
```
- It seems that this function has not been committed
- Is it really a general utility? It makes assumptions of the inner structure of the qubit object. I would suggest to have a method qb.get_channel_map() since only the qb object should know which of its parameters store channel names that are relevant for a plot. We could still consider having a wrapper that combines the channel maps of individual qubits to a multi-qubit channel map. This wrapper might belong into the device object.

chellings replied:
First comment would be addressed by cherry-picking as you suggested (yes, it would be OK to cherry-pick), but what about the second comment? Also note that the function you intend to cherry-pick has no docstring.

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+            qubits = self.meas_objs
+        plot_kwargs.update(dict(channel_map=plot_kwargs.pop('channel_map',
+                           general.get_channel_map(qubits))))
+        plot_kwargs.update(dict(legend=plot_kwargs.pop('legend',False)))
```
Is it really a good idea to have no legend by default?

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        segments by default.
+        :param qubits (list): list of qubits to plot
+        :param save (bool): save the figures.
+        :param plot_kwargs: kwargs passed on to segment.plot()
```
maybe indicate which keys of plot_kwargs get modified before passing it on to segment.plot()

nathlacroix replied:
done

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        if sequences is None:
+            # plot all sequences
+            sequences = self.sequences
+        elif isinstance(sequences, int):
```
Not sure whether that is needed, but it should not be difficult to also allow a single string.
```
else:
    if not isinstance(sequences, list):
        sequences = [sequences]
    # the logic that processes the list goes here
```

nathlacroix replied:
done

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+            if len(sequences) and isinstance(sequences[0], int):
+                # sequences is a list of indices of sequences to plot
+                sequences = [self.sequences[s] for s in sequences]
+            elif len(sequences) and isinstance(sequences[0], str):
```
Taking the decision based on the [0] element if not what is reported in the docstring ("If strings are in the list").

What about allowing hybrid lists? (maybe not needed, but also not difficult)
```
sequences = [[s for i, s in enumerate(self.sequences) if i == ind or s.name == ind] for ind in sequences]
```

nathlacroix replied:
done

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+
+        for seq, segs in zip(sequences, segments):
+            for s in segs:
+                if isinstance(s, int):
```
what is the alternative? does this need an else? not sure I understand the logic here

chellings replied:
Comment not addressed. Please add a comment in the code.

Do I see correctly that you have not included Kuno's bugfix? Is it no longer needed?

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,74 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+                                       bbox_inches="tight")
+                    plot_kwargs.update(dict(save_kwargs=save_kwargs,
+                                            savefig=True))
+                seq.segments[s].plot(**plot_kwargs)
```
You should collect the return values (if any, i.e., if the user sets show_and_close=False). Either return them in a list or store them as a property?

nathlacroix replied:
I think at least returning them makes sense. I added that. With respect of making them a property, it is likely not essential but I could see the benefit of doing it. From an object oriented perspective, I think it would make more sense to attach it to the segment (e.g. seg.fig) rather than the QExp. What do you think?

chellings reviewed (changes requested):
Thanks for the new feature. As you pointed out to me, this is still in kind of a draft stage, so I indeed have some suggestions for improvements. No hurry to get it improved and merged, but I think it is good that the suggested features and the suggested changes are now documented already. I can re-review once you had time to work on it.

nathlacroix committed:
Add plot sequence saving to qexp and default plotting before the measurement.

nathlacroix committed:
Add channel_map function to utilities.general to be able to access it within pycqed
Could be discussed whether the assumptions made about the channel map semantics
are too specific. One could in a later stage modify the function to allow as input
names of the qcode parameters to look for for each qubit.

Commit:
QE.plot: bugfix in indexing of segment
np.arange is not of type int, but np.int32

nathlacroix committed:
Improve QuantumExperiment plotting function

- Complement doc string
- More flexibility for hybrid list of sequence names and indices
- return figures and axes

nathlacroix commented:
I pushed a commit addressing your comments, @chellings . The one open item that remains for me is the default behavior. after using this function quite a bit I realize I typically want to plot the first segment by default and not all of them. What do you think?
On another note: I notice this PR requires commit 496389957c1baa738debce603d5197239406c0a4 for the channel map. is that commit intended to be used in another (preceding) PR? If not, I would cherry pick it on this branch, even though it is not directly related to plotting.

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -670,33 +670,38 @@ def plot(self, sequences=None, segments=None, qubits=None,
+            indices must be provided: the outer list corresponds to each
+            sequence and the inner list to the indices of the segments to plot.
+            E.g. segments=[[0,1],[3]] will plot the 0th and 1rst segment of
+                sequence 0 and the 3d segment of sequence 1.
```
typos: 1rst and 3d

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -705,29 +710,31 @@ def plot(self, sequences=None, segments=None, qubits=None,
-                seq.segments[s].plot(**plot_kwargs)
-
+                figs_and_axs.append(seq.segments[s].plot(**plot_kwargs))
+        return figs_and_axs
```
```
        # avoid returning a list of Nones (if show_and_close is True)
        return [v for v in figs_and_axs if v is not None] or None
```

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -661,5 +662,81 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+            E.g. segments=[[0,1],[3]] will plot the 0th and 1rst segment of
+                sequence 0 and the 3d segment of sequence 1.
+            Plots all segments by default.
+        :param qubits (list): list of qubits to plot. Defaults to self.meas_objs
```
comment still open:
- please clarify whether list of objects or list of strings (names). Consider passing it to self.get_qubits to allow the user passing qb names.

chellings reviewed (changes requested):
Thanks for the improvements. It seems that some of my comments have been overlooked, see the unresolved conversations and the new comments. If you intentionally do not implement a comment, please write a short response. Otherwise I assume in the re-review that it has been overlooked.

I do not have a strong opinion on the default behavior (plotting all segements or only the first). Feel free to adapt it in the way you find most useful. Thank you!

About returning or storing the plots: storing them in seg sounds interesting, but let's just move on now with returning them, keeping the possibility of storing them in seg in mind as a possibility for the future.

nathlacroix committed:
Add doc string to get_channel_map

Refactor parameter "ge" to more explicit name "drive"

nathlacroix committed:
Changes to Qexp.plot

- Change default behavior to plotting first segment and first sequence only instead of all segments and all sequences
- Allow to pass qubit names
- Do not return figures if they are not returned by segment.plot
- Improve doc string

nathlacroix commented:
Thanks @chellings for the comments.
I have included the changes and cherrypicked the commits of interest

nathlacroix commented:
@chellings I have also added now the commit we had said should also go in this PR. Note that they created little merge conflicts which I think I have taken care of properly. Let me know if you notice something strange

chellings commented on pycqed/measurement/waveform_control/segment.py:
```diff
@@ -754,7 +754,7 @@ def get_segment_start_end(self):
         return np.min(start_end_times[:, 0]), np.max(start_end_times[:, 1])
 
-    def _test_overlap(self):
+    def _test_overlap(self, allow_overlap=False, tol=1e-12):
```
It seems that you cherry-picked an improvement for a feature that does not exist on the master. allow_overlap is part of MR !3. It won't work on this branch because it also needs changes in other functions.

I anyways do not see the relation of this to the current PR. It rather seems to be related to MR !3, and I can cherry-pick it there.

nathlacroix committed:
Resolve conflict in qexp

nathlacroix commented:
Sorry for the confusion, I got confused by the fact that MR !3 was already in the master. I forced-pushed to reset the situation to a reasonable status. @chellings does that seem fine for you now?

nathlacroix committed:
Improve segment plotting formatting

- remove "no_codeword" from legend entry
- align ylabels
- remove frame from legend

nathlacroix committed:
Move channel map method from general to qubit and device objects

nathlacroix committed:
Use channel map from device if available in qexp.plot

nathlacroix committed:
Change default behavior of legend in qexp.plot.

Also add the legend argument to the signature to allow for autocomplete

chellings commented on pycqed/measurement/quantum_experiment.py:
```diff
@@ -735,5 +739,89 @@ def save_timers(self, quantum_experiment=True, sequence=True, segments=True, fil
+        qubits, _ = self.get_qubits(qubits) # get qubit objects
+        default_ch_map = \
+            self.dev.get_channel_map(qubits) if self.dev is not None else \
+                {qb.name: qb.get_channel_map() for qb in qubits}
```
Shouldn't this be as follows, or am I missing something?
```
{qb.name: qb.get_channel_map()[qb.name] for qb in qubits}
```

nathlacroix replied:
There was indeed a typo. I intended to use qubit.get_channels() which returns already a list without the need to pop the key.

chellings reviewed (approved):
Please check my last small comment (potential bug?). Assuming that you have tested the final changes, I approve. Thanks for the feature and all the fine-tuning during the review process.

nathlacroix committed:
Fix call to get_channels instead of get_channel_map() in QExp.plot
